Add beginnings of quota support for emptyDir on XFS.

dgoodwin opened the pull request:
Wraps the k8s empty_dir volume plugin in one which layers on code to apply
quotas if the volume dir is on XFS, mounted with gquota, and an FSGroup is
available for the pod.

This is likely a stepping stone to a more permanent solution in the future.

More details on the research into this proposed approach are available here: https://github.com/openshift/origin/pull/7206

This attempt makes an effort to make no kubernetes modifications we need to carry. Instead the goal was to accomplish this purely in origin and look to push to something upstream in time.

Alternatively we could just modify the empty_dir plugin in kubernetes and carry the patches for time being.

dgoodwin commented:
Likely of interest to @ironcladlou and @smarterclayton 

Still some debugging code in there as I'm working on a couple things still, the node-config.yaml setting doesn't seem to be getting read for one.

ironcladlou commented:
>Still some debugging code in there as I'm working on a couple things still, the node-config.yaml setting doesn't seem to be getting read for one. 

You added the field to the internal types, but you also need it to the public versioned types (e.g. v1), and then regenerate conversions, etc.

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,128 @@
+		return nil
+	}
+	if !xqa.isXfs(dir) {
+		glog.V(3).Infof("%s is not on an XFS filesystem, skipping quota application.")
```
Seems like maybe this needs to be smarter, otherwise emptyDir on a misconfigured node would allow all pods to use unrestricted storage. I would expect failure to apply quota to propagate back up as a failure to establish the volume for the pod (except if the pod is somehow special and allowed to bypass quota application).

dgoodwin replied:
Hmm yes I suppose so, originally I was thinking it's a plugin that's in play regardless, but if there's an explicit config option to turn this on then failure to apply the quota does seem pretty bad. I'll rig this up, I think I'll skip wrapping the plugin entirely unless you have something in your config file for it, and then treat it like a proper error internally.

smarterclayton replied:
Quota right now is going to be admin policy applied to the node.  If
the policy is enabled, failure to apply the policy should definitely
be an error.  When it moves to the pod, if the node can't satisfy the
policy, that should also be an error.

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+// Check if the empty dir is on an XFS filesystem:
+func (xqa *xfsQuotaApplicator) isXfs(dir string) bool {
```
I prefer to return (bool, error) so the caller can distinguish "yes" vs. "no" vs. "I don't know because of an error", but don't feel too strongly.

smarterclayton replied:
Always return errors

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,128 @@
+		volDevice,
+	}
+
+	// xfs_quota is very happy to fail but return a success code, likely due to it's
```
s/it's/its

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,128 @@
+	// xfs_quota is very happy to fail but return a success code, likely due to it's
+	// interactive shell approach. Grab stderr, if we see anything written to it we'll
+	// consider this an error.
+	cmd := exec.Command("xfs_quota", args...)
```
Is this the accepted way to execute shell commands from infra code? I haven't yet tried to do it, so I wonder if there are existing utils/abstractions. One concern is the possibility of the call to `Run()` blocking forever.

smarterclayton replied:
There should at least be shims for testing.

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,128 @@
+// run df on the emptydir path, returning a header line and the line we're
+// interested in. The first string token in that line will be the device name.
+func getFsDevice(dir string) (string, error) {
+	outBytes, err := exec.Command("df", dir).Output()
```
If you do `df --output=source $dir` you can make stronger assumptions about the output

ironcladlou commented on pkg/cmd/server/api/types.go:
```diff
@@ -116,6 +116,11 @@ type NodeConfig struct {
+	// ProjectEmptyDirQuota is the quota applied to emptyDir volumes specified in MB.
+	// This is currently implemented only for XFS, and is a limit per namespace per node.
+	// (this may change in the future)
+	ProjectEmptyDirQuota int
```
Should this be a pointer? Otherwise you'll need some magic number as the "no quota" value.

dgoodwin commented:
Ok I think this is ready for review, dropping the WIP tag. I will squash commits once closer to merge but leaving them separate for now in case reverts are needed. Since original PR was opened:

- fixed the reading of projectEmptyDirQuota from node-config.xml.
- using an int * for the NodeConfig API type to support the value not being set, and yet allow the use of 0 effectively shutting down emptyDir usage.
- failure to apply the quota is now treated as an error
- use df --output=source for more reliable lookup of the device for a dir
- added a suite of tests for quota.go.
- uppercased some abbreviations
- added tests for loading node-config.xml files with various projectEmptyDirQuota settings, or none at all.
- added validation code to prevent negative projectEmptyDirQuota settings, and associated tests.

Is it feasible to think about end to end integration tests here? It would require the runtime environment to be on XFS with group quota enabled and tests running as root. If so any advice on a starting point (how to make sure this is the case, or who to contact) would be appreciated.

ironcladlou commented on pkg/cmd/server/api/validation/node_test.go:
```diff
@@ -61,3 +62,23 @@ func TestFailingKubeletArgs(t *testing.T) {
+	if emptyDirQuotaError == nil {
+		t.Fatalf("expected projectEmptyDirQuota error but got none")
+	}
+	if !strings.Contains("must be a positive integer", emptyDirQuotaError.Detail) {
```
Could also check for `errors.IsInvalid`

ironcladlou commented on pkg/volume/empty_dir/empty_dir_quota.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+// Wrapper for the k8s empty dir plugin builder:
+type EmptyDirQuotaPluginBuilder struct {
```
Applies throughout: GoDoc convention is "// <symbol name> <words>"

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,162 @@
+func (xqa *xfsQuotaApplicator) isXFS(dir string) (bool, error) {
+	out, _, err := xqa.cmdRunner.RunFsTypeCommand(dir)
+	if err != nil {
+		glog.Errorf("error checking filesystem type for %s: %s", dir, err)
```
Capitalization

ironcladlou commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,162 @@
+		return "", errors.New("insufficient lines in df output")
+	}
+
+	fsDevice := strings.Split(lines[1], " ")[0]
```
This could panic if the line count is as expected but the line content is malformed; better do a bounds check

dgoodwin replied:
I think this might be correct as is, lines[1] should be safe via the check above, and splitting any string should always return a [0] as far as I can tell. Added some more tests to try to make sure but let me know if I missed something in this.

dgoodwin commented:
Thanks for the quick review @ironcladlou, I think I've got everything updated now.

ironcladlou commented:
This looks really close to me; let's find a second reviewer.

dgoodwin commented:
Awesome. I will squash it down to one commit now.

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -116,6 +116,11 @@ type NodeConfig struct {
+	// ProjectEmptyDirQuota is the quota applied to emptyDir volumes specified in MB.
+	// This is currently implemented only for XFS, and is a limit per namespace per node.
+	// (this may change in the future)
+	ProjectEmptyDirQuota *int
```
I haven't followed this, but why would I want a different way to express `resource.Quantity`?

dgoodwin replied:
@deads2k apologies I'm not quite familiar with that, could you (or anyone else) fill me in on what that's a ref to?

deads2k replied:
> @deads2k apologies I'm not quite familiar with that, could you (or anyone else) fill me in on what that's a ref to?

When specifying limits for things that take `MB` or other standard units in the API, we've used this type: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/resource/quantity.go#L91.  I would expect this configuration file to be consistent in its types unless there is a strong reason to deviate.

dgoodwin replied:
Great idea I will see if I can make it work.

dgoodwin commented:
Ok updated to use resource.Quantity throughout and added some more testing. Had to register the conversion func in the server/api/v1/conversions.go.

ironcladlou commented on pkg/cmd/server/start/node_args.go:
```diff
@@ -147,6 +148,10 @@ func (args NodeArgs) BuildSerializeableNodeConfig() (*configapi.NodeConfig, erro
 
+	// TODO: Remove this:
+	tempDefaultQuota := resource.MustParse("1024Gi")
+	internal.(*configapi.NodeConfig).ProjectEmptyDirQuota = &tempDefaultQuota
```
Did you intend to commit this?

dgoodwin replied:
Nope my bad. Fixed now.

ironcladlou commented:
Refactored API LGTM, @deads2k, want to give it another look?

deads2k commented:
I haven't looked at the implementation just the api.  Did @liggitt approve adding the top level element to the config?  That looks off to my eye.

dgoodwin commented:
What do you think @liggitt ? 

The top level node config in question is:

projectEmptyDirQuota: 512Mi

It's probably not a permanent config option as we hope this will eventually fold into upstream quota support.

deads2k commented:
> It's probably not a permanent config option as we hope this will eventually fold into upstream quota support.

Once its in this API, we aren't going to want to rip it out.  People will count on this for protection.

I'm not very close to the area, but the idea of a `VolumeConfig` stanza doesn't seem out of place.  Won't nodes have to be able to say which kinds of volumes they support?

dgoodwin commented:
@deads2k if I understood @smarterclayton correctly it's not that the functionality will go away just that it'll move, change, and improve. Rather than a blanket node-config setting I'd hope it eventually becomes something applied per project/namespace for better granularity, likely in the quota subsystem. (at least that's where it felt best)

However if it does live on I'm certainly not opposed to improving where the config lands. How does this look:

volumeConfig:
  projectEmptyDirQuota: 512Mi

Or option (2):

volumeConfig:
  emptyDir:
    projectQuota: 512Mi

liggitt commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -68,6 +69,11 @@ type NodeConfig struct {
+	// ProjectEmptyDirQuota is the resource quantity to be used as quota for emptyDir volumes.
+	// This is currently implemented only for XFS, and is a limit per namespace per node.
+	// (this may change in the future)
+	ProjectEmptyDirQuota *resource.Quantity `json:"projectEmptyDirQuota"`
```
the upstream types aren't versioned?

dgoodwin replied:
Upstream type being resource.Quantity? I don't think it's versioned, only hit I can find is in Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/resource/quantity.go and there don't appear to be any version sub-dirs there. It serializes to just a string so possibly deemed not required for this.

liggitt commented on pkg/cmd/server/kubernetes/node_config.go:
```diff
@@ -169,6 +171,32 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
+		wrappedEmptyDirPlugin := false
+		for idx, plugin := range cfg.VolumePlugins {
+			// Can't really do type checking or use a constant here as they are not exported:
+			if plugin.Name() == "kubernetes.io/empty-dir" {
```
I'm really uncomfortable with this... I'd rather carry a patch to the upstream type that makes it satisfy an interface that declares a `SetDefaultQuota` function.

dgoodwin replied:
@pmorie cautioned me strongly at the outset against modifying the actual emptyDir implementation in k8s, so the plan was to avoid that if at all possible (which it turned out it was), and check in with @pmorie about how we could adapt it to the actual kube API in the future.

However @pmorie later softened on that stance and if we're sure, I can re-do the plugin code and push it into a patch we carry on top of kube.

However to make the case for the current implementation, it offers a lower maintenance burden and will hopefully one day move to a permanent API location. It relies on volume plugin builder's NewBuilder method, and the volume plugin's SetUp and SetUpAt. I don't know how stable these methods are, I would hope somewhat, but indeed they could change upstream. 

Please let me know if we're sure and I'll get to work on moving everything into kube.

ironcladlou replied:
@smarterclayton please weight in on this

deads2k replied:
Carrying a patch will warn us if this changed during a rebase.  Without that interlock, this looks ripe for slipping.

ironcladlou replied:
Does it really seem likely the plugin interface methods in use here are going to change during a minor or other rebase without us noticing? I agree that a carried patch to emptyDir is technically cleaner, I'm wondering if there really is enough risk to warrant reworking at this point.

smarterclayton replied:
If this goes upstream, will we patch empty dir or expose a new volume
plugin?

On Feb 25, 2016, at 4:13 PM, Dan Mace <notifications@github.com> wrote:

In pkg/cmd/server/kubernetes/node_config.go
<https://github.com/openshift/origin/pull/7351#discussion_r54165366>:

> @@ -169,6 +171,32 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
>  		return nil, err
>  	}
>
> +	// Replace the standard k8s emptyDir volume plugin with a wrapper version
> +	// which offers XFS quota functionality, but only if the node config
> +	// specifies an empty dir quota to apply to projects:
> +	if options.ProjectEmptyDirQuota != nil {
> +		glog.V(2).Info("Replacing empty-dir volume plugin with quota wrapper")
> +		wrappedEmptyDirPlugin := false
> +		for idx, plugin := range cfg.VolumePlugins {
> +			// Can't really do type checking or use a constant here as they are not exported:
> +			if plugin.Name() == "kubernetes.io/empty-dir" {

Does it really seem likely the plugin interface methods in use here are
going to change during a minor or other rebase without us noticing? I agree
that a carried patch to emptyDir is technically cleaner, I'm wondering if
there really is enough risk to warrant reworking at this point.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/7351/files#r54165366>.

dgoodwin replied:
I don't think it would be a new volume plugin, we want basically exactly what emptyDir does just with some added functionality, so long term I'd like to see it embedded within k8s emptyDir plugin or volume plugins in general. Doing so as a new plugin would likely be a bunch of copied code that has to be manually maintained.

deads2k replied:
> Does it really seem likely the plugin interface methods in use here are going to change during a minor or other rebase without us noticing?

Yes.  `RESTMapper` broke two weeks ago and it is used in every client command that we run.  It manifested as some broken shortcuts which were only caught by our e2e test.  I don't see corresponding tests here, so that would result in a slip and effectively disabling this feature silently.  And by and large, rebasers know more about `RESTMappers` than `VolumePlugins`.

deads2k replied:
> If this goes upstream, will we patch empty dir or expose a new volume
> plugin?

This is still an emptyDir, it just supports an optional size limit.

ironcladlou replied:
Remove the inlining and use delegating calls to the emptydir plugin. If those methods were removed from emptyDir (which I still contend is something we would know about), we'd get compiler errors.

dgoodwin replied:
How do folks feel about this as a compromise solution? ^^ This saves us having to go back and reworking and re-testing (as we're hoping to get this in for online launch), avoids the rebase breakage problem, and as ever we can hopefully take a crack and adapting it directly into the upstream API later.

ironcladlou replied:
And `var _ volume.VolumePlugin = &emptyDirQuota{}`

deads2k replied:
Why not test what you actually need to match: `CanSupport`?  Matching on a magic, non-guaranteed name seems less than ideal and you want to become the owner for a particular kind of volume source, not an override for a bad impl.

If you don't find a match at all, is there a reason why you shouldn't add yourself to the list to handle emptDirs?

dgoodwin replied:
@deads2k @liggitt you guys ok with the composition rather than inheritance compromise above? Should eliminate the silent breakage during a rebase concern. Going to start work on it momentarily if so.

Re: CanSupport, sounds good I will try that out.

This plugin can't do much without an instance of an emptyDir plugin that can actually do the work, so we'd need to instantiate one ourselves if it's not already in the list. Afaik you can't configure the list of plugins to be enabled today, so if emptyDir were to not be there we probably want to be looking into why, it might be another situation where we should fail fast. I will take a look though and see.

deads2k replied:
> @deads2k @liggitt you guys ok with the composition rather than inheritance compromise above?

I'm ok with it, though this is making more work later.  You should get @markturansky to help build a way to configure `VolumePlugins`.  Which ones are available, how to configure settings on them, and better organizing which cases they handle.

liggitt replied:
be careful with CanSupport... I vaguely remember the emptydir plugin also handling secret volumes, but I might be misremembering

dgoodwin replied:
Excellent thanks. 

Mark actually helped me figure out how to do this without a k8s patch before I started off, and mentioned we were looking at a way to do just that. (configure volume plugins) I was tempted to do so as part of this but it was a bit too far out of scope.

markturansky replied:
@deads2k the binary process that starts up kubelet and the controllermanager is the part that knows what the volume plugin list is.  Currently, Origin looks to the upstream binary for the list, which is all plugins.  We can configure our own list. Suggestions on where this would go?

deads2k replied:
> @deads2k the binary process that starts up kubelet and the controllermanager is the part that knows what the volume plugin list is. Currently, Origin looks to the upstream binary for the list, which is all plugins. We can configure our own list. Suggestions on where this would go?

Does upstream not have interest in allowing cluster-admins to completely disable some volume plugins or in having configurable settings on them like admission plugin?  I'd prefer to make use of upstream infrastructure and simply do the wiring in origin.

liggitt commented on pkg/volume/empty_dir/empty_dir_quota.go:
```diff
@@ -0,0 +1,66 @@
+
+// EmptyDirQuotaPluginBuilder is a simple wrapper for the k8s empty dir plugin builder.
+type EmptyDirQuotaPluginBuilder struct {
+	volume.VolumePlugin
```
inlining upstream types is a good way to silently break when an interface changes and the functions you thought you were overriding no longer get called

dgoodwin commented:
A little more backstory, we're under a bit of time crunch and the situation upstream is in flux in terms of disk usage quota, so decision was to solve what we could for now and adapt it over time, this is why we didn't immediately tackle full quota sub-system implementation and/or integrating directly into Kube API initially.

deads2k commented:
> However if it does live on I'm certainly not opposed to improving where the config lands. How does this look:
> 
> volumeConfig:
> projectEmptyDirQuota: 512Mi
> 
> Or option (2):
> 
> volumeConfig:
> emptyDir:
> projectQuota: 512Mi

If you know of other knobs you already want for different volume types, then use option 2.  Otherwise, start with

```yaml
volumeConfig:
  emptyDirProjectQuota: 512Mi
```

Movement of fields within the lower types is more narrowly scoped if we end up needing to do it.

liggitt commented:
Is the limit per namespace, or per fsgroup? If I'm reading correctly, it limits non-memory, XFS, per-node, per-fsgroup quota. Not sure what to call the config setting that does that

smarterclayton commented:
LocalVolumeQuota struct {
  // if specified enforces quota on empty died per unique fsgroup
  FSGroup *int64
}

dgoodwin commented:
Looks good to me, I will combine with @deads2k idea and go with:

volumeConfig:
  localVolumeQuota:
    fsGroup: 512Mi

Let me know if anyone has issues with this.

liggitt commented:
`localQuota`... volume is implied
would we use this setting across volume types, or only for empty dir? if only empty dir, that should probably  still be indicated in the setting

dgoodwin commented:
Ack re LocalQuota. 

Certainly just emptyDir for now, looking through the list of volume types, I can't really see much else *today* that it could apply to, doesn't feel like a good idea for hostPath, it would be nice if it kicked in for the autocreated volumes when you've got a container with a VOLUME that isn't assigned someday though.

The local designation feels like a good idea for future proofing to me.

dgoodwin commented:
Individual commits added to help with review (will squash later), I think this covers everything discussed.

Travis is giving me attitude over swagger, will look into that next.

dgoodwin commented:
Added docs on the container API types. Squashed everything into one commit, rebased on latest master this morning.

Travis is still complaining about swagger and I do not understand the output, it's showing my new additions to the API as *removed* in the diff. hack/update-generated-swagger-spec.sh does nothing locally even after rebasing on top of latest master. Can anyone clarify what's going on there or if there's something I need to do?

dgoodwin commented:
I think I found this, (a) learned you must keep local master branch in sync with upstream/master (I've just been merging that into whatever branch I'm working on), and (b) I was running hack/update-generated-swagger-spec.sh, not -descriptions.sh.

Branch updated, should make Travis happy.

dgoodwin commented:
Right now if you attempt to use this but the filesystem isn't XFS, the error surfaces when pods try to use emptyDir volumes. Would it make sense to error out early during node startup if you try to do this and your volume directory is not on XFS?

dgoodwin commented:
Added an additional commit that will cause the node to fail to start if you request a local quota, but the volume directory is not on a supported filesystem type.

Also added better support for adding more filesystem types in the future, you request a quota applicator now, the code checks the type and returns an appropriate implementation if it's supported, otherwise an error. (right now only XFS is supported)

dgoodwin commented:
[test]

dgoodwin commented:
[test]

dgoodwin commented:
Squashed into one commit, tests are flaking but will keep trying, @deads2k if you've got time, how is this looking on your end? Hoping to make 3.2 for online.

dgoodwin commented:
[test]

dgoodwin commented:
![687474703a2f2f73332e616d617a6f6e6177732e636f6d2f72617067656e6975732f313239313133313638305f74776f2d7468756d62732d75702e6a7067](https://cloud.githubusercontent.com/assets/51265/13471013/2bb0d010-e085-11e5-89e4-781a277810eb.jpg)

dgoodwin commented:
@deads2k @liggitt anybody have time to take another look? Tests now cleared.

smarterclayton commented:
Needs an extended test.

dgoodwin commented:
@smarterclayton will start on that this morning, what's the timeline for tagging what will roll to online? I have some PTO booked next week, wondering if we could take this as is if the current code is ok, and I could submit extended test as a followup PR?

smarterclayton commented:
If you can't finish the test today, we'll get someone on the team to help
finish it.

On Mar 4, 2016, at 7:56 AM, Devan Goodwin <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> will start on that this
morning, what's the timeline for tagging what will roll to online? I have
some PTO booked next week, wondering if we could take this as is if the
current code is ok, and I could submit extended test as a followup PR?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/7351#issuecomment-192271181>.

markturansky commented:
@dgoodwin I volunteer to help.

dgoodwin commented:
@markturansky you are a brave soul, thanks!

 I will either get it done today, shuffle the PTO to later in week, or send you details on where it's at.

dgoodwin commented:
I've been talking with @danmcp about what this will involve on the testing infra / images side. The requirements for this test to pass will be the volume directory on an XFS filesystem, mounted with gquota option. This will not pass out of the box today as the base images currently define the logical volumes, and they're not set up for this. 

If we proceed, Dan proposes we modify our approach to get this a bit better controlled.

1. Update the base images (created by hand, just to the point of getting vagrant deps involved) so they define the volume groups, but not the logical volumes. 
  * Dan says rebuilding all these images would take a half day if you know what you're doing. 
1. Update vagrant-openshift to define the logical volumes so going forward we have better control over things like this. 
  * We'll now create / mount LV's as we wish in either install_origin_base_dependencies.rb or a new similar step.

So this will be a bit involved and is going to take some time before the dust settles. Please speak up if anyone feels this is too involved, otherwise I will march on with writing the test and having it pass locally, then focus on how we get the images/builds ready for it.

I do propose that we consider taking this PR as is so we know it's in for online, and adding the extended test as a follow-up PR. Will leave that up to you guys. 

CC @abhgupta @ironcladlou  ^^ for your thoughts on that. (not sure how much time is left)

dgoodwin commented:
@smarterclayton and anyone else who sees this, I could use a sanity check on what I'm having to do to write this test, it's going to be quite tightly coupled with vagrant-openshift, you won't be able to just run it on your own system without some manual setup as it needs to make pretty big assumptions about the system.

We need voldir on an XFS filesystem, and technically we need to trash that filesystem before or after the test runs as there's no easy way to delete an XFS quota once applied, at the very least we'd have to unmount it. Do the extended tests run on a fresh vm every time or does the vm ever get re-used?

@danmcp added room in a volume group for the base images for me to create a new small LV, it'll be called something like openshift-xfs-vol-dir and mounted to /tmp/openshift/xfs-vol-dir. I intend to unmount and format this before each run.

The test does not fit the structure of any existing extended tests I can find, it requires a node-config change so it will need a separate launcher shell script to modify node-config and then kick off the tests, which will be written in go.

If any of the above sounds off please let me know ASAP. It's not a pretty test and not something I'm used to writing, so I could use some guidance on what'll be acceptable during review before the fact if possible.

dgoodwin commented:
Rebased against master and a new commit added with the extended test. This will likely not pass in jenkins yet as we're still sorting out some issues with getting the logical volume in place that the test will need, but hopefully sometime today.

Test uses a separate launcher script as we need a node-config.yaml change, it is adapted from core.sh with as much unnecessary cruft removed as I dared.

dgoodwin commented:
Ready for review @deads2k @smarterclayton @ironcladlou if anyone is free!

smarterclayton commented:
Why do we need the separate launcher script?  I would just run this in the
default configuration and have it use XFS quota by default.

On Thu, Mar 10, 2016 at 9:28 AM, Devan Goodwin <notifications@github.com>
wrote:

> Ready for review @deads2k <https://github.com/deads2k> @smarterclayton
> <https://github.com/smarterclayton> @ironcladlou
> <https://github.com/ironcladlou> if anyone is free!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7351#issuecomment-194871549>.
>

dgoodwin commented:
At present kubelet won't come up if you specify a quota, but you're not on XFS, I felt like it was best treated like a config error and you need to address it to run the node.

As such if I plug this into core.sh as the default behavior, nobody can run any extended test locally without doing the manual setup for this. (xfs partition mounted to /tmp/openshift/xfs-vol-dir with gquota) Seemed like too much of a hindrance to all future devs given most people are not going to be on XFS by default.

smarterclayton commented:
Detect whether /tmp/openshift/xfs-vol-dir exists and turn the flag on if it
does.  We *want* this to be part of every automated test.  I'm ok with
developers not testing it if Jenknis is.

On Thu, Mar 10, 2016 at 10:14 AM, Devan Goodwin <notifications@github.com>
wrote:

> At present kubelet won't come up if you specify a quota, but you're not on
> XFS, I felt like it was best treated like a config error and you need to
> address it to run the node.
>
> As such if I plug this into core.sh as the default behavior, nobody can
> run any extended test locally without doing the manual setup for this. (xfs
> partition mounted to /tmp/openshift/xfs-vol-dir with gquota) Seemed like
> too much of a hindrance to all future devs given most people are not going
> to be on XFS by default.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7351#issuecomment-194897840>.
>

dgoodwin commented:
Good idea, will make that change ASAP.

deads2k commented on pkg/cmd/server/api/validation/node.go:
```diff
@@ -113,3 +115,14 @@ func ValidateDockerConfig(config api.DockerConfig, fldPath *field.Path) field.Er
+func ValidateVolumeConfig(config api.VolumeConfig, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	fsGroupLocalQuotaPath := fldPath.Child("localQuota").Child("fsGroup")
```
`fldPath.Child("localQuota", "fsGroup")`.  Also, why not embed it right where you use it below?

deads2k commented on pkg/volume/empty_dir/empty_dir_quota.go:
```diff
@@ -0,0 +1,104 @@
+	Wrapped volume.VolumePlugin
+
+	// The default quota to apply to each node:
+	DefaultQuota resource.Quantity
```
Is there something else that can override this "Default" quota?  I'm not seeing it.

dgoodwin replied:
No, it was planned but scrapped during development, I will rename that to just Quota.

deads2k commented on pkg/volume/empty_dir/empty_dir_quota.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func (plugin *EmptyDirQuotaPlugin) Name() string {
+	return plugin.Wrapped.Name()
```
Why do we lie about our name?

dgoodwin replied:
If you change the plugin name there's pretty significant breakage, I believe with any pre-existing emptyDir volume. Kube I believe is somehow mapping the name to the volume metadata and thus we'd hit major upgrade problems rolling this out. 

Let me know if this is a serious concern and I can dig deeper.

markturansky replied:
I am fairly certain the volume plugin name is only used by the pluginManager to maintain a distinct list.  Volumes can be obtained by name from the manager, but they are generally found using the `CanSupport` method, which in this case would return true when volumeSource.EmptyDir is not nil.

@dgoodwin do we call Init for each plugins before or after we swap out the emptyDir plugin for our wrapped plugin?

If we init after, then the plugin name from *our* plugin will be used in the volumeManager.  No one outside the plugin packaging knows or cares what plugin names are.

deads2k replied:
> Let me know if this is a serious concern and I can dig deeper.

I'm not super concerned.  I just like being able to tell that different things are different.

dgoodwin replied:
I took another quick attempt at this, I think the problem is in the volume cleaner somewhere @markturansky, If I jump around between master and my branch creating pods with empty dir volumes I start seeing issues like:

E0310 15:46:01.836059   22555 volumes.go:242] Could not create volume cleaner for jenkins-1-centos7-volume-1: can't use volume plugins for 0772ba24-e6f8-11e5-931d-989096de63cb/kubernetes.io~empty-dir: no volume plugin matched

It looked like rolling out a rename was more trouble than it would be worth.

deads2k replied:
> It looked like rolling out a rename was more trouble than it would be worth.

ok

markturansky replied:
You're right.  Here is the plugin name encoded in the filepath so that a plugin can later be created by name:

https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/empty_dir/empty_dir.go#L328

deads2k commented on pkg/volume/empty_dir/empty_dir_quota.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func (edq *emptyDirQuotaBuilder) GetMetrics() (*volume.Metrics, error) {
+	return edq.wrapped.GetMetrics()
```
Do you need to perform `min` operations for `Capacity` and `Available`?

dgoodwin replied:
I'm not super familiar with this stuff but I believe we were considering capacity / metrics as a possible future enhancement in the design: https://github.com/openshift/origin/pull/7206

Do I have that right @ironcladlou

ironcladlou replied:
I'm not sure I understand the question re: `min`. Re: the proposal, I can confirm that reporting or otherwise dealing with metrics used to drive scheduling is considering out of scope for this PR. Does that answer the question?

deads2k replied:
> I'm not sure I understand the question re: min. Re: the proposal, I can confirm that reporting or otherwise dealing with metrics used to drive scheduling is considering out of scope for this PR. Does that answer the question?

If quota is 100MB and capacity is 500MB, the actual capacity is 100MB right?  Again, I don't fully understand all the repercussions, but it seems inconsistent.

deads2k commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,183 @@
+// quotaCommandRunner interface is used to abstract the actual running of
+// commands so we can unit test more behavior.
+type quotaCommandRunner interface {
+	RunFsTypeCommand(dir string) (string, string, error)
```
Abbreviations are all the same case: `FS`.

deads2k commented on pkg/volume/empty_dir/quota.go:
```diff
@@ -0,0 +1,183 @@
+	return fsDevice, parseErr
+}
+
+func ParseFsDevice(dfOutput string) (string, error) {
```
godoc why this is public.  It looks very specific.

dgoodwin replied:
Hmm, this was so I could use it in my extended test, but looking at it I probably should expose something higher level.

deads2k replied:
> Hmm, this was so I could use it in my extended test, but looking at it I probably should expose something higher level.

I'm not requiring a refactor, just some explicit docs that warn people off from using it would satisfy me.

dgoodwin commented:
With exception of the plugin name, all modifications are now made.

dgoodwin commented:
@smarterclayton @deads2k I think everything is addressed from yesterday, let me know if there's any more to be done.

dgoodwin commented:
@smarterclayton PR updated:

- Lengthened 4 lines (made 2 passes through, those were all I could spot)
- Fixed godoc mismatch.
- Deduplicated XFS stat output checking by pushing to a common method in the package.
- Check if stat returned XFS by comparing to direct string instead of Contains, after trimming whitespace, which is what I was really worried about.
- Bumped all log levels in quota.go.
- Stopped logging errors in quota.go, return the data in the error itself instead.
- Make sure xfs_quota is on PATH before we create the XFS quota applicator. Node will not start if this is not the case.

Squashed into one commit.

pmorie commented on pkg/cmd/server/api/types.go:
```diff
@@ -116,6 +117,23 @@ type NodeConfig struct {
+	// FSGroup can be specified to enable a quota on local storage use per unique FSGroup ID.
+	// At present this is only implemented for emptyDir volumes, and if the underlying
+	// volumeDirectory is on an XFS filesystem.
+	FSGroup *resource.Quantity
```
why is this a resource.Quantity instead of an *int64?

dgoodwin replied:
It was but this was requested way above by David, it lets us use things like 256Mi / 2Gi in the node-config.yaml setting, keeps everything more consistent with the other places we use Quantity.

pmorie replied:
Oh, I see, this isn't an FSGroup itself, but the quota per fsgroup -- I feel like the name might be confusing, how about 'PerFSGroup'

dgoodwin replied:
That does seem clearer, I will adjust.

dgoodwin commented:
Updated per feedback from @smarterclayton and @pmorie.

smarterclayton commented:
Can you run the extended as part of this PR yet (via test extended)?

dgoodwin commented:
If I'm following your question, yes it should be included in a default extended test suite run as part of this PR: https://github.com/dgoodwin/origin/blob/quota-emptydir/test/extended/extended_test.go#L12

It will fail if you don't have proper mount point setup, which our images should have at this point.

smarterclayton commented:
Then [testonlyextended][extended:core(local storage quota)]

On Mon, Mar 14, 2016 at 12:18 PM, Devan Goodwin
<notifications@github.com> wrote:
> If I'm following your question, yes it should be included in a default
> extended test suite run as part of this PR:
> https://github.com/dgoodwin/origin/blob/quota-emptydir/test/extended/extended_test.go#L12
>
> It will fail if you don't have proper mount point setup, which our images
> should have at this point.
>
> —
> Reply to this email directly or view it on GitHub.

smarterclayton commented:
Looks like it failed

On Mon, Mar 14, 2016 at 1:35 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/testonlyextended FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2086/)
> (Extended Tests: core(local storage quota))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7351#issuecomment-196426431>.
>

dgoodwin commented:
Forgive me while I abuse this PR to get some debug info. 

[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
This might do it.

[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
Trying to figure out what to do here, we setup the base images to mount an XFS filesystem to /tmp/openshift/xfs-vol-dir, but it would appear something is later coming along and mounting /tmp as a tmpfs, which looks like it's effectively hiding the /tmp/openshift/xfs-vol-dir mount point and preventing the test from being able to use it and pass.

smarterclayton commented:
@dgoodwin, that's something we have to do on the test machines.  @deads2k
we may have a problem here where we have to change the ordering of the /tmp
mount.

On Tue, Mar 15, 2016 at 9:15 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2131/)
> (Extended Tests: core(local storage quota))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/7351#issuecomment-196812074
>

dgoodwin commented:
@smarterclayton I think I found a good route past it, I've modified the images to mount to /mnt/openshift-xfs-vol-dir instead, and I have an update to the test here locally that I'll push as soon as the images are ready. This avoids any problems with tmp and it can continue to be mounted as it is today.

I'll push the change as soon as they images are done building, that is underway as we speak.

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
This latest failure is looking more like a legit test failure and getting further now that the images have all been rebuilt, it's possible it's because I only gave it 30 seconds for the pod to create and quota be applied. Adding more debug info and increasing timeout for another try.

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)] 

![georgemichaelcollapse](https://cloud.githubusercontent.com/assets/51265/13814550/79a3b428-eb65-11e5-9ee6-577b2b6f23ba.gif)

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
[testonlyextended][extended:core(local storage quota)]

dgoodwin commented:
Extended test is now passing, there was a bug in the test logic if run against a fresh XFS partition, most of my work here was reusing a partition due to the complications with "deleting" an XFS quota. 

Should be ready for merge.

smarterclayton commented:
Please squash and remove all debugging logic.

dgoodwin commented:
Apologies forgot all about the debug commits. Updated now.

dgoodwin commented:
All green. Hoping to make Friday's build if possible!

smarterclayton commented:
[merge]

On Wed, Mar 16, 2016 at 2:45 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test SUCCESS (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2206/)
> (Extended Tests: core(local storage quota))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7351#issuecomment-197481804>
>

dgoodwin commented:
Flake, re[test] please.

ironcladlou commented:
:joy:
